### PR TITLE
added support for node range

### PR DESCRIPTION
**Pull request by NickyDaB**
# Added Support for node range

## Purpose
- working on https://github.com/NickyDaB/CAST/issues/15
- working on https://github.com/IBM/CAST/issues/222
- to give cmd line wrappers for CSM APIs support for a node range syntax like `node[01-09]`

## Approach
- not exactly a full solution to the core issue, but I did some work on this feature a few weeks ago in the R&D branch when I was looking into the issue with allocation query active all.
- I cleaned up some of that functionality into a function call.
- probably easy to use in multiple APIs.
- node attributes update and query is an example of using new function tool

#### Open Questions and Pre-Merge TODOs
- [ ] @pdlun92 please add a test case for this new feature
- Open Question: I'm not sure if we should bundle in `-n` like I did here, or expand out into a seperate flag like `-N, --node_range` if we do the second, then I can print out details of why the tool failed. but if we stick to the `-n` then I can't get specific because there may not be a node range on purpose ie: `-n "node01"` but keeping everything in `-n` seems more straight forward for a user. @whowutwut agrees to keep it as `-n`. look at `~/utilities/src/string_tools.c"` for more info

## How to Test

1. test the new feature. 
2. the tool itself can return multiple error codes that can be checked by a caller to report details. 
3. cmd line interfaces for `node_attributes_query` and `node_attributes_update` are the only two apis that use this tool at the moment. 

## Screenshots

```
[root@c650f03p41 bin]# ./csm_cluster_query_state 
---
# node_name  | collection_time            | update_time                | state      | type    | num_allocs - allocs - states - shared
#            +                            +                            +            +         +  
# c650f03p41 | 2018-08-27 13:59:30.917032 | 2018-08-27 13:58:32.653634 | DISCOVERED | utility |  0
# n01        | 1999-01-27 01:00:00        | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 1 - running - f
# n02        | 2016-01-27 02:00:00        | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 1 - running - f
# n03        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 1 - running - f
# n04        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 2 - running - f
# n05        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 2 - running - f
# n06        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 2 - running - f
# n07        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 3 - running - f
# n08        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 3 - running - f
# n09        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  1 - 3 - running - f
# n10        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n30        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n31        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n32        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n33        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n34        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n35        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
...
```

```
./csm_node_attributes_update -n n[01-09] -s "IN_SERVICE"
```

```
[root@c650f03p41 bin]# ./csm_cluster_query_state 
---
# node_name  | collection_time            | update_time                | state      | type    | num_allocs - allocs - states - shared
#            +                            +                            +            +         +  
# c650f03p41 | 2018-08-27 13:59:30.917032 | 2018-08-27 13:58:32.653634 | DISCOVERED | utility |  0
# n01        | 1999-01-27 01:00:00        | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 1 - running - f
# n02        | 2016-01-27 02:00:00        | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 1 - running - f
# n03        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 1 - running - f
# n04        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 2 - running - f
# n05        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 2 - running - f
# n06        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 2 - running - f
# n07        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 3 - running - f
# n08        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 3 - running - f
# n09        | 2018-08-27 13:58:03.465555 | 2018-08-27 16:55:20.115871 | IN_SERVICE | compute |  1 - 3 - running - f
# n10        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n30        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n31        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n32        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n33        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n34        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
# n35        | 2018-08-27 13:58:03.465555 | 2018-08-27 13:58:03.465555 | DISCOVERED | compute |  0
...

```

### CODE REVIEW

File: utilities/src/string_tools.c [hunk 1]
```diff
@@ -249,6 +249,201 @@ int CORAL_stringTools_nodeCount_xCATSyntax(char* myString, int* dataCount)
...
+		
+		//record the number of characters in the string before the '['
+		//aka, the base node name
+		reserved_chars = bracket_pointer-myString;
+
+		//find the dash
+		char* dash_pointer = NULL;
+		dash_pointer = strchr(myString,'-');
+
+		if(dash_pointer == NULL)
```
Comment (mew2057) [inline utilities/src/string_tools.c]: Future iterations should allow something like `c650f02p[5]`, as xCAT allows this syntax.

File: utilities/src/string_tools.c [hunk 2]
```diff
@@ -249,6 +249,201 @@ int CORAL_stringTools_nodeCount_xCATSyntax(char* myString, int* dataCount)
...
+		char* first_node_digit = NULL;
+
+		first_node_digit = (char*)calloc(range_digits+1,sizeof(char));
+		//memcpy the stuff out of myString
+		memcpy(first_node_digit, bracket_pointer+1, range_digits);
+
+		//grab the first node digit. save to int. 
+		//use this later to find total number of nodes
+		int first_node_digit_int = 0;
+		first_node_digit_int = atoi(first_node_digit);
```
Comment (mew2057) [inline utilities/src/string_tools.c]: I think there's a bug here. I think this range: [10-200] will result in the first character being retrieved as `10-`. The atoi will then fail. The `range_digits` name is a bit hard to follow.

Comment (NickyDaB) [reply utilities/src/string_tools.c]: range_digits would come out to be 2 for the '10' and 3 for the '200'. its a number calculated by subtracting pointer locations. at lines 327, 345. (i reuse it, store the other values into `_begin` and `_end`) 

eventually on line 359, i create a temp string to copy the chars (but I just realized now I never freed it. gotta go back and update that) using `range_digits` to determine how many chars to copy. 

so now after 363 `first_node_digit` is a string that contains `10` (in your example) - (but it also wouldn't get this far because 10 is less precise in digits than 200.) that check is done on 349 . 

then once `first_node_digit` has data, we convert `first_node_digit` into `first_node_digit_int` vai `atoi`.

I will update the comment to:
//number of digits in the node range, calculated by subtracting pointer locations

after lunch we can talk about my comments and how I can make them more straight forward, or where they didn't provide enough detail.

File: utilities/src/string_tools.c (hunk 1 above)
Comment (NickyDaB) [reply utilities/src/string_tools.c]: I think for CSM 2 we might want to not use this tool at all, and prep for using the xCAT apis for getting node ranges and stuff or something, idk, mark and vic mentioned it when i was asking about stuff. 

My thinking here is that there are too many combinations to worry about via reinventing the wheel. there are many examples, but one is brackets in the middel of a node name, ie: `c123f[02-30]p05` or excluding node names ie: `c123f01p[01-10]-c123f01p07` or however that syntax is, but i know a feature like that is supported.

**Comment by NickyDaB** [comment]
@mew2057 

> I want to see a test run where a range has two numbers with a different number of digits (e.g. 1-100) to make sure there's no glaring bugs.

this would result in the function completing with a return code of 3. explained on line 267. Did I give bad wording as to suggest something else? maybe i should add the word `precision` in there or something?

**Comment by mew2057** [comment]
@NickyDaB I think the help should reflect this in outlining how the range finder requires same width numbers.

**Comment by NickyDaB** [comment]
@mew2057 

> I think the help should reflect this in outlining how the range finder requires same width numbers, just as a point of clarity. But if that's the response the API will have I'll approve it.

There is no help, but there is a function declaration comment. Is that what you are referring to?

**Comment by NickyDaB** [comment]
@mew2057 , @williammorrison2 , @pdlun92 

I pushed new commits to this pull request. I have the "protected" branch set up that if this happens, then all previous approved reviews must be re approved. So you will notice that your "green checks" have gone back to "un reviewed". I think this is proper procedure, because these new changes can potentially change the original approved code.
